File: app/rust/src/ch03_immutable_data.rs
//! 第3章: イミュータブルなデータ操作
//!
//! Rust ではデフォルトで不変（イミュータブル）です。
//! Vec のスライス操作と新しいコレクションの作成を学びます。

// =============================================================================
// 3.1 イミュータブルとは
// =============================================================================

/// Rust では `let` で宣言した変数はデフォルトで不変
/// 変更するには新しい値を作成する
///
/// # Examples
///
/// ```
/// let list = vec!["Apple", "Book"];
/// // list.push("Mango"); // コンパイルエラー！
///
/// // 新しい Vec を作成
/// let mut new_list = list.clone();
/// new_list.push("Mango");
///
/// assert_eq!(list.len(), 2);     // 元のリストは変わらない
/// assert_eq!(new_list.len(), 3); // 新しいリストに追加された
/// ```
pub fn immutable_demo() {
    let list = vec!["Apple", "Book"];
    let mut new_list = list.clone();
    new_list.push("Mango");
    assert_eq!(list.len(), 2);
    assert_eq!(new_list.len(), 3);
}

// =============================================================================
// 3.2 Vec の基本操作
// =============================================================================

/// 要素を追加した新しい Vec を返す（元の Vec は変更しない）
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::appended;
///
/// let apple_book = vec!["Apple", "Book"];
/// let apple_book_mango = appended(&apple_book, "Mango");
///
/// assert_eq!(apple_book.len(), 2);      // 元のリストは変わらない
/// assert_eq!(apple_book_mango.len(), 3);
/// assert_eq!(apple_book_mango, vec!["Apple", "Book", "Mango"]);
/// ```
pub fn appended<T: Clone>(list: &[T], element: T) -> Vec<T> {
    let mut result = list.to_vec();
    result.push(element);
    result
}

/// 2つのスライスを結合した新しい Vec を返す
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::appended_all;
///
/// let ab = vec!["a", "b"];
/// let cd = vec!["c", "d"];
/// let abcd = appended_all(&ab, &cd);
///
/// assert_eq!(abcd, vec!["a", "b", "c", "d"]);
/// assert_eq!(ab.len(), 2); // 元のリストは変わらない
/// ```
pub fn appended_all<T: Clone>(list: &[T], other: &[T]) -> Vec<T> {
    let mut result = list.to_vec();
    result.extend_from_slice(other);
    result
}

/// 最初の2要素を取得
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::first_two;
///
/// assert_eq!(first_two(&["a", "b", "c"]), vec!["a", "b"]);
/// assert_eq!(first_two(&["x", "y"]), vec!["x", "y"]);
/// assert_eq!(first_two(&["z"]), vec!["z"]);
/// ```
pub fn first_two<T: Clone>(list: &[T]) -> Vec<T> {
    list.iter().take(2).cloned().collect()
}

/// 最後の2要素を取得
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::last_two;
///
/// assert_eq!(last_two(&["a", "b", "c"]), vec!["b", "c"]);
/// assert_eq!(last_two(&["x", "y"]), vec!["x", "y"]);
/// assert_eq!(last_two(&["z"]), vec!["z"]);
/// ```
pub fn last_two<T: Clone>(list: &[T]) -> Vec<T> {
    let len = list.len();
    if len <= 2 {
        list.to_vec()
    } else {
        list[len - 2..].to_vec()
    }
}

// =============================================================================
// 3.3 リストの変換
// =============================================================================

/// 最初の2要素を末尾に移動
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::move_first_two_to_end;
///
/// assert_eq!(
///     move_first_two_to_end(&["a", "b", "c"]),
///     vec!["c", "a", "b"]
/// );
/// assert_eq!(
///     move_first_two_to_end(&["a", "b", "c", "d"]),
///     vec!["c", "d", "a", "b"]
/// );
/// ```
pub fn move_first_two_to_end<T: Clone>(list: &[T]) -> Vec<T> {
    if list.len() <= 2 {
        return list.to_vec();
    }
    let first_two = &list[..2];
    let without_first_two = &list[2..];
    appended_all(without_first_two, first_two)
}

/// 最後の要素の前に新しい要素を挿入
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::insert_before_last;
///
/// assert_eq!(
///     insert_before_last(&["a", "b"], "c"),
///     vec!["a", "c", "b"]
/// );
/// assert_eq!(
///     insert_before_last(&["x"], "y"),
///     vec!["y", "x"]
/// );
/// ```
pub fn insert_before_last<T: Clone>(list: &[T], element: T) -> Vec<T> {
    if list.is_empty() {
        return vec![element];
    }
    let len = list.len();
    let without_last = &list[..len - 1];
    let last = &list[len - 1..];
    let mut result = without_last.to_vec();
    result.push(element);
    result.extend_from_slice(last);
    result
}

// =============================================================================
// 3.4 旅程の再計画
// =============================================================================

/// 指定した都市の前に新しい都市を挿入
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::replan;
///
/// let plan_a = vec!["Paris", "Berlin", "Kraków"];
/// let plan_b = replan(&plan_a, "Vienna", "Kraków");
///
/// assert_eq!(plan_b, vec!["Paris", "Berlin", "Vienna", "Kraków"]);
/// assert_eq!(plan_a, vec!["Paris", "Berlin", "Kraków"]); // 元の計画は変わらない
/// ```
pub fn replan<'a>(plan: &[&'a str], new_city: &'a str, before_city: &str) -> Vec<&'a str> {
    let before_city_index = plan.iter().position(|&c| c == before_city);

    match before_city_index {
        Some(index) => {
            let cities_before = &plan[..index];
            let cities_after = &plan[index..];
            let mut result = cities_before.to_vec();
            result.push(new_city);
            result.extend_from_slice(cities_after);
            result
        }
        None => {
            // 指定した都市が見つからない場合は末尾に追加
            let mut result = plan.to_vec();
            result.push(new_city);
            result
        }
    }
}

/// ジェネリック版の旅程再計画
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::replan_generic;
///
/// let plan: Vec<String> = vec!["Paris".into(), "Berlin".into(), "Kraków".into()];
/// let krakow = "Kraków".to_string();
/// let plan_b = replan_generic(&plan, "Vienna".into(), &krakow);
///
/// assert_eq!(plan_b, vec!["Paris", "Berlin", "Vienna", "Kraków"]);
/// ```
pub fn replan_generic<T: Clone + PartialEq>(plan: &[T], new_city: T, before_city: &T) -> Vec<T> {
    let before_city_index = plan.iter().position(|c| c == before_city);

    match before_city_index {
        Some(index) => {
            let mut result = plan[..index].to_vec();
            result.push(new_city);
            result.extend_from_slice(&plan[index..]);
            result
        }
        None => {
            let mut result = plan.to_vec();
            result.push(new_city);
            result
        }
    }
}

// =============================================================================
// 3.5 String とスライスの類似性
// =============================================================================

/// 名前を省略形にする
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::abbreviate;
///
/// assert_eq!(abbreviate("Alonzo Church"), "A. Church");
/// assert_eq!(abbreviate("Alan Turing"), "A. Turing");
/// ```
pub fn abbreviate(name: &str) -> String {
    if let Some(space_index) = name.find(' ') {
        let initial = &name[..1];
        let last_name = &name[space_index + 1..];
        format!("{}. {}", initial, last_name)
    } else {
        name.to_string()
    }
}

/// 文字列の部分を取得
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::substring;
///
/// assert_eq!(substring("hello", 1, 4), "ell");
/// assert_eq!(substring("Rust", 0, 2), "Ru");
/// ```
pub fn substring(s: &str, start: usize, end: usize) -> &str {
    &s[start..end]
}

// =============================================================================
// 3.6 中央への挿入
// =============================================================================

/// リストの中央に要素を挿入
///
/// # Examples
///
/// ```
/// use grokking_fp::ch03_immutable_data::insert_at_middle;
///
/// assert_eq!(
///     insert_at_middle(&["a", "b", "c", "d"], "X"),
///     vec!["a", "b", "X", "c", "d"]
/// );
/// assert_eq!(
///     insert_at_middle(&["a", "b"], "X"),
///     vec!["a", "X", "b"]
/// );
/// ```
pub fn insert_at_middle<T: Clone>(list: &[T], element: T) -> Vec<T> {
    let middle = list.len() / 2;
    let before = &list[..middle];
    let after = &list[middle..];
    let mut result = before.to_vec();
    result.push(element);
    result.extend_from_slice(after);
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_appended() {
        let list = vec![1, 2, 3];
        let result = appended(&list, 4);
        assert_eq!(result, vec![1, 2, 3, 4]);
        assert_eq!(list, vec![1, 2, 3]); // 元のリストは変わらない
    }

    #[test]
    fn test_appended_all() {
        let list1 = vec![1, 2];
        let list2 = vec![3, 4];
        let result = appended_all(&list1, &list2);
        assert_eq!(result, vec![1, 2, 3, 4]);
    }

    #[test]
    fn test_first_two() {
        assert_eq!(first_two(&[1, 2, 3, 4, 5]), vec![1, 2]);
        assert_eq!(first_two(&[1]), vec![1]);
        assert_eq!(first_two::<i32>(&[]), Vec::<i32>::new());
    }

    #[test]
    fn test_last_two() {
        assert_eq!(last_two(&[1, 2, 3, 4, 5]), vec![4, 5]);
        assert_eq!(last_two(&[1]), vec![1]);
        assert_eq!(last_two::<i32>(&[]), Vec::<i32>::new());
    }

    #[test]
    fn test_move_first_two_to_end() {
        assert_eq!(move_first_two_to_end(&[1, 2, 3, 4]), vec![3, 4, 1, 2]);
        assert_eq!(move_first_two_to_end(&[1, 2]), vec![1, 2]);
        assert_eq!(move_first_two_to_end(&[1]), vec![1]);
    }

    #[test]
    fn test_insert_before_last() {
        assert_eq!(insert_before_last(&[1, 2, 3], 99), vec![1, 2, 99, 3]);
        assert_eq!(insert_before_last(&[1], 99), vec![99, 1]);
        assert_eq!(insert_before_last::<i32>(&[], 99), vec![99]);
    }

    #[test]
    fn test_replan() {
        let plan = vec!["Paris", "Berlin", "Kraków"];
        let new_plan = replan(&plan, "Vienna", "Kraków");
        assert_eq!(new_plan, vec!["Paris", "Berlin", "Vienna", "Kraków"]);
        assert_eq!(plan, vec!["Paris", "Berlin", "Kraków"]); // 元は変わらない
    }

    #[test]
    fn test_replan_city_not_found() {
        let plan = vec!["Paris", "Berlin"];
        let new_plan = replan(&plan, "Vienna", "NotExist");
        assert_eq!(new_plan, vec!["Paris", "Berlin", "Vienna"]); // 末尾に追加
    }

    #[test]
    fn test_abbreviate() {
        assert_eq!(abbreviate("Alonzo Church"), "A. Church");
        assert_eq!(abbreviate("SingleName"), "SingleName");
    }

    #[test]
    fn test_insert_at_middle() {
        assert_eq!(insert_at_middle(&[1, 2, 3, 4], 99), vec![1, 2, 99, 3, 4]);
        assert_eq!(insert_at_middle(&[1, 2, 3], 99), vec![1, 99, 2, 3]);
        assert_eq!(insert_at_middle::<i32>(&[], 99), vec![99]);
    }
}

File: app/typescript/src/ch03_immutable_data.ts
/**
 * 第3章: イミュータブルなデータ操作
 *
 * イミュータブル（不変）なデータの操作方法を学びます。
 * データを「変更」する代わりに、新しいデータを「作成」します。
 */

import { pipe } from 'fp-ts/function'
import * as RA from 'fp-ts/ReadonlyArray'

// =============================================================================
// 3.1 イミュータブルなリスト操作
// =============================================================================

/**
 * リストに要素を追加した新しいリストを返す。
 * 元のリストは変更されない。
 *
 * @example
 * const original = ["Apple", "Book"] as const
 * const newList = appended(original, "Mango")
 * // original: ["Apple", "Book"]
 * // newList: ["Apple", "Book", "Mango"]
 */
export const appended = <T>(
  lst: readonly T[],
  element: T
): readonly T[] => [...lst, element]

/**
 * リストに複数の要素を追加した新しいリストを返す。
 *
 * @example
 * appendedAll(["a", "b"], ["c", "d"]) // => ["a", "b", "c", "d"]
 */
export const appendedAll = <T>(
  lst: readonly T[],
  elements: readonly T[]
): readonly T[] => [...lst, ...elements]

/**
 * リストの先頭に要素を追加した新しいリストを返す。
 *
 * @example
 * prepended(["b", "c"], "a") // => ["a", "b", "c"]
 */
export const prepended = <T>(
  lst: readonly T[],
  element: T
): readonly T[] => [element, ...lst]

// =============================================================================
// 3.2 スライス操作
// =============================================================================

/**
 * リストの最初の n 要素を取得する。
 *
 * @example
 * firstN(["a", "b", "c"], 2) // => ["a", "b"]
 * firstN(["a"], 2) // => ["a"]
 */
export const firstN = <T>(lst: readonly T[], n: number): readonly T[] =>
  lst.slice(0, n)

/**
 * リストの最初の2要素を取得する。
 *
 * @example
 * firstTwo(["a", "b", "c"]) // => ["a", "b"]
 */
export const firstTwo = <T>(lst: readonly T[]): readonly T[] =>
  firstN(lst, 2)

/**
 * リストの最後の n 要素を取得する。
 *
 * @example
 * lastN(["a", "b", "c"], 2) // => ["b", "c"]
 * lastN(["a"], 2) // => ["a"]
 */
export const lastN = <T>(lst: readonly T[], n: number): readonly T[] =>
  lst.length < n ? [...lst] : lst.slice(-n)

/**
 * リストの最後の2要素を取得する。
 *
 * @example
 * lastTwo(["a", "b", "c"]) // => ["b", "c"]
 */
export const lastTwo = <T>(lst: readonly T[]): readonly T[] =>
  lastN(lst, 2)

/**
 * 最初の2要素を末尾に移動した新しいリストを返す。
 *
 * @example
 * moveFirstTwoToEnd(["a", "b", "c"]) // => ["c", "a", "b"]
 * moveFirstTwoToEnd(["a", "b", "c", "d"]) // => ["c", "d", "a", "b"]
 */
export const moveFirstTwoToEnd = <T>(lst: readonly T[]): readonly T[] => {
  const first = lst.slice(0, 2)
  const rest = lst.slice(2)
  return [...rest, ...first]
}

/**
 * 最後の要素の前に要素を挿入した新しいリストを返す。
 *
 * @example
 * insertBeforeLast(["a", "b"], "c") // => ["a", "c", "b"]
 * insertBeforeLast(["a"], "b") // => ["b", "a"]
 */
export const insertBeforeLast = <T>(
  lst: readonly T[],
  element: T
): readonly T[] => {
  if (lst.length === 0) return [element]
  return [...lst.slice(0, -1), element, ...lst.slice(-1)]
}

/**
 * リストの中央に要素を挿入した新しいリストを返す。
 *
 * @example
 * insertAtMiddle(["a", "b", "c", "d"], "X") // => ["a", "b", "X", "c", "d"]
 * insertAtMiddle(["a", "b"], "X") // => ["a", "X", "b"]
 */
export const insertAtMiddle = <T>(
  lst: readonly T[],
  element: T
): readonly T[] => {
  const middle = Math.floor(lst.length / 2)
  return [...lst.slice(0, middle), element, ...lst.slice(middle)]
}

// =============================================================================
// 3.3 旅程の再計画
// =============================================================================

/**
 * 指定した都市の前に新しい都市を挿入した旅程を返す。
 *
 * @example
 * const plan = ["Paris", "Berlin", "Kraków"]
 * replan(plan, "Vienna", "Kraków") // => ["Paris", "Berlin", "Vienna", "Kraków"]
 * // plan は変わらない: ["Paris", "Berlin", "Kraków"]
 */
export const replan = (
  plan: readonly string[],
  newCity: string,
  beforeCity: string
): readonly string[] => {
  const index = plan.indexOf(beforeCity)
  if (index === -1) return [...plan, newCity]
  return [...plan.slice(0, index), newCity, ...plan.slice(index)]
}

// =============================================================================
// 3.4 文字列操作（String もイミュータブル）
// =============================================================================

/**
 * 名前を省略形にする。
 *
 * @example
 * abbreviate("Alonzo Church") // => "A. Church"
 * abbreviate("A. Church") // => "A. Church"
 */
export const abbreviate = (name: string): string => {
  const separator = name.indexOf(' ')
  if (separator === -1) return name
  const initial = name.charAt(0)
  const lastName = name.substring(separator + 1)
  return `${initial}. ${lastName}`
}

/**
 * 部分文字列を取得する。
 *
 * @example
 * substring("hello", 1, 4) // => "ell"
 * substring("TypeScript", 0, 4) // => "Type"
 */
export const substring = (s: string, start: number, end: number): string =>
  s.substring(start, end)

// =============================================================================
// 3.5 イミュータブルなオブジェクト
// =============================================================================

/** 2次元の点 */
export interface Point {
  readonly x: number
  readonly y: number
}

/** Point を作成する */
export const createPoint = (x: number, y: number): Point => ({ x, y })

/** Point の x 座標を更新した新しい Point を返す */
export const withX = (point: Point, newX: number): Point => ({
  ...point,
  x: newX,
})

/** Point の y 座標を更新した新しい Point を返す */
export const withY = (point: Point, newY: number): Point => ({
  ...point,
  y: newY,
})

/** 都市 */
export interface City {
  readonly name: string
  readonly population: number
}

/** City を作成する */
export const createCity = (name: string, population: number): City => ({
  name,
  population,
})

/** 人口を更新した新しい City を返す */
export const withPopulation = (city: City, newPopulation: number): City => ({
  ...city,
  population: newPopulation,
})

// =============================================================================
// 3.6 ReadonlySet（イミュータブルな集合）
// =============================================================================

/**
 * Set に要素を追加した新しい Set を返す。
 *
 * @example
 * const s = new Set(["a", "b"])
 * addToSet(s, "c") // => Set(["a", "b", "c"])
 * // s は変わらない
 */
export const addToSet = <T>(s: ReadonlySet<T>, element: T): ReadonlySet<T> =>
  new Set([...s, element])

/**
 * Set から要素を削除した新しい Set を返す。
 *
 * @example
 * const s = new Set(["a", "b", "c"])
 * removeFromSet(s, "b") // => Set(["a", "c"])
 */
export const removeFromSet = <T>(s: ReadonlySet<T>, element: T): ReadonlySet<T> =>
  new Set([...s].filter((e) => e !== element))

// =============================================================================
// 3.7 ReadonlyMap（イミュータブルな辞書）
// =============================================================================

/**
 * Map のキーを更新した新しい Map を返す。
 *
 * @example
 * const m = new Map([["a", 1], ["b", 2]])
 * updateMap(m, "c", 3) // => Map([["a", 1], ["b", 2], ["c", 3]])
 */
export const updateMap = <K, V>(
  m: ReadonlyMap<K, V>,
  key: K,
  value: V
): ReadonlyMap<K, V> => new Map([...m, [key, value]])

/**
 * Map からキーを削除した新しい Map を返す。
 *
 * @example
 * const m = new Map([["a", 1], ["b", 2], ["c", 3]])
 * removeKey(m, "b") // => Map([["a", 1], ["c", 3]])
 */
export const removeKey = <K, V>(
  m: ReadonlyMap<K, V>,
  key: K
): ReadonlyMap<K, V> => new Map([...m].filter(([k]) => k !== key))

// =============================================================================
// 3.8 fp-ts を使ったイミュータブル操作
// =============================================================================

/**
 * fp-ts の ReadonlyArray を使って要素を追加する。
 *
 * @example
 * appendWithFpts(["a", "b"], "c") // => ["a", "b", "c"]
 */
export const appendWithFpts = <T>(
  lst: readonly T[],
  element: T
): readonly T[] =>
  pipe(lst, RA.append(element))

/**
 * fp-ts の ReadonlyArray を使って先頭に要素を追加する。
 *
 * @example
 * prependWithFpts(["b", "c"], "a") // => ["a", "b", "c"]
 */
export const prependWithFpts = <T>(
  lst: readonly T[],
  element: T
): readonly T[] =>
  pipe(lst, RA.prepend(element))

/**
 * fp-ts の ReadonlyArray を使ってリストを連結する。
 *
 * @example
 * concatWithFpts(["a", "b"], ["c", "d"]) // => ["a", "b", "c", "d"]
 */
export const concatWithFpts = <T>(
  lst1: readonly T[],
  lst2: readonly T[]
): readonly T[] =>
  pipe(lst1, RA.concat(lst2))
